### PR TITLE
[FIX] website_sale: optimize controller of /shop

### PR DESCRIPTION
**Pull request by rco-odoo**
When the shop has many products, the controller makes pathological
queries to retrieve the categories and attributes of the products
matching the search criteria.  The patch consists in composing the final
queries with the query to retrieve the products, instead of injecting
the ids of the products.

In the example database, the shop has around 400K products, and the page
`/shop` takes 18 seconds to render.  The optimization reduces the page
rendering to 2 seconds.

### CODE REVIEW

**Comment by JKE-be** [comment]
If i'm not wrong, the goal of Odoo (at least of CTO) is not to support  big ecommerce.
So do we really need to make this code/controller so un-orm-able ?
We could probably improve a lot of method, if we bybass orm with direct queries 

It will not improve the shop of 3 essential oils of @antonylesuisse, but next time that Anthony will open the code, he will delete all and say 'what is this crap'. no?

File: addons/website_sale/controllers/main.py [hunk 1]
```diff
@@ -234,10 +234,36 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
...
+            Category._apply_ir_rules(query, 'read')
+            cat_from, cat_where, cat_params = query.get_sql()
+
+            field = type(Product).public_categ_ids
+
+            query = """
+                SELECT "{cat_table}".id FROM {cat_from}
+                WHERE {cat_where} AND "{cat_table}".id IN (
+                    SELECT rel.{rel_cat} FROM {rel_table} rel
+                    WHERE rel.{rel_prod} IN ({prod_query})
```
Comment (Yajo) [inline addons/website_sale/controllers/main.py]: Adding a SELECT in a WHERE of a SELECT that is in a WHERE is possibly the worst way to optimize things. Try doing it with JOINs wherever possible.

File: addons/website_sale/controllers/main.py [hunk 2]
```diff
@@ -251,14 +277,66 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
...
+                        SELECT "{attline_table}".attribute_id FROM {attline_from}
+                        WHERE {attline_where} AND "{attline_table}".id IN (
+                            SELECT "{rel_attline}" FROM "{rel_table}"
+                        )
+                    )
+                    AND "{att_table}".id IN (
+                        SELECT "{attline_table}".attribute_id FROM {attline_from}
+                        WHERE {attline_where}
+                            AND "{attline_table}".product_tmpl_id IN ({prod_query})
+                    )
```
Comment (Yajo) [inline addons/website_sale/controllers/main.py]: Again try to avoid subqueries to ensure fastest results.

**Comment by antonylesuisse** [comment]
Yes. If we find that the orm is an order of magnitude slower we need to figure exactly why why and fix it instead of bypassing it.

**Comment by wtaferner** [comment]
@antonylesuisse 
Isn't it fixed in 13.0? :wink:

**Comment by leemannd** [comment]
Hello @JKE-be , @rco-odoo ,
Thanks for your work, when I'm reading `is not to support big ecommerce.`, I'm just wondering what is your actual target in terms of product catalog size.

**Comment by rco-odoo** [comment]
This was just a POC, and gave us some directions to follow for future developments.

**Pull request closed without merging**